cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,179 @@
+            return Status::Invalid("Empty " +
+                                   static_cast<std::string>(kSessionCookieName) +
+                                   " cookie value.");
+        }
+      }
+      if (!session_id.empty()) break;
+    }
+
+    if (session_id.empty()) {
+      // No cookie was found
+      *middleware = std::shared_ptr<ServerSessionMiddleware>(
+          new ServerSessionMiddleware(this, incoming_headers));
+    } else {
+      try {
+        const std::shared_lock<std::shared_mutex> l(session_store_lock_);
+        auto session = session_store_.at(session_id);
+        *middleware = std::shared_ptr<ServerSessionMiddleware>(
+            new ServerSessionMiddleware(this, incoming_headers, session, session_id));
+      } catch (std::out_of_range& e) {
+        return Status::Invalid("Invalid or expired " +
+                               static_cast<std::string>(kSessionCookieName) + " cookie.");
+      }
+    }
+
+    return Status::OK();
+  }
+
+  /// \brief Get a new, empty session option map and its id key.
+  std::shared_ptr<FlightSqlSession> GetNewSession(std::string* session_id) {
+    std::string new_id = boost::lexical_cast<std::string>(uuid_generator_());
+    *session_id = new_id;
+    auto session = std::make_shared<FlightSqlSession>();
+
+    const std::unique_lock<std::shared_mutex> l(session_store_lock_);
+    session_store_[new_id] = session;
+
+    return session;
+  }
+};
+
+ServerSessionMiddleware::ServerSessionMiddleware(ServerSessionMiddlewareFactory* factory,
+                                                 const CallHeaders& headers)
+    : factory_(factory), headers_(headers), existing_session(false) {}
+
+ServerSessionMiddleware::ServerSessionMiddleware(
+    ServerSessionMiddlewareFactory* factory, const CallHeaders& headers,
+    std::shared_ptr<FlightSqlSession> session, std::string session_id)
+    : factory_(factory),
+      headers_(headers),
+      session_(std::move(session)),
+      session_id_(std::move(session_id)),
+      existing_session(true) {}
+
+void ServerSessionMiddleware::SendingHeaders(AddCallHeaders* addCallHeaders) {
+  if (!existing_session && session_) {
+    addCallHeaders->AddHeader(
+        "set-cookie", static_cast<std::string>(kSessionCookieName) + "=" + session_id_);
+  }
+}
+
+void ServerSessionMiddleware::CallCompleted(const Status&) {}
+
+bool ServerSessionMiddleware::HasSession() const { return static_cast<bool>(session_); }
+
+std::shared_ptr<FlightSqlSession> ServerSessionMiddleware::GetSession() {
+  if (!session_) session_ = factory_->GetNewSession(&session_id_);
+  return session_;
+}
+
+const CallHeaders& ServerSessionMiddleware::GetCallHeaders() const { return headers_; }
+
+std::shared_ptr<ServerMiddlewareFactory> MakeServerSessionMiddlewareFactory() {
+  return std::shared_ptr<ServerSessionMiddlewareFactory>(
+      new ServerSessionMiddlewareFactory());
+}
+
+::arrow::Result<SessionOptionValue> FlightSqlSession::GetSessionOption(
```

Comment:
Might be semantically clearer to use `std::optional` if the only possible error is that the key might not be found

Reply:
Agreed; done.